### DISCUSSION TITLE
Is it possible to use one View with multiple ViewModels?

### DISCUSSION THREAD

**Discussion by Gigas002** [AvaloniaUI/Avalonia#6530] [Q&A] [answered]
I wonder if there’s a possibility to create a single `View` with multiple `ViewModels`.
Let’s suppose I have a `UserControl` with `Button` in it. And this control refers to a single item in `TabControl`. So the views are completely the same, but behavior of button’s logic can drastically change. And I have like 10 different views for this purpose, though I could use just one.
I understand, that I could create a single `View` with single `ViewModel` and just create a `switch` inside of buttons to check for selected `TabItem`, and then run the needed part of programming logic, but I find this solution less beautiful, than theoretical usage of multiple `ViewModels`.

**Comment by jp2masa** [comment, marked as answer]
I think you're mixing some concepts, 10 different views refers to view instances (i.e. 10 `<MyView />`) or view types (i.e. 10 different user control XAMLs)?

If you want to change the view model for a view you can just bind the `DataContext` and change it in the parent view model when needed:

```xaml
<DemoView DataContext="{Binding Demo}" />
```

If you're using `TabControl`, you can just use the same view type for all items, with different view models:

```xaml
<TabControl>
    <TabItem Header="First">
        <TabView DataContext="{Binding FirstTab}" />
    </TabItem>
    <TabItem Header="Second">
        <TabView DataContext="{Binding SecondTab}" />
    </TabItem>
</TabControl>
```

**Reply by maxkatz6** [reply]
Also in some cases it's more convenient to have array of view models (using some base type/interface with Header information). And use this array as TabControl items source. In ItemTemplate you can decide on specific template depending on the view model type.

**Reply by Gigas002** [reply]
I'm binding my views in `MainWindowView` to `TabControl` the following way:

```xaml
    <TabControl Items="{Binding Tabs}">
      <TabControl.ItemTemplate>
        <DataTemplate>
          <TextBlock Text="{Binding Header}"/>
        </DataTemplate>
      </TabControl.ItemTemplate>
      <TabControl.ContentTemplate>
        <DataTemplate DataType="vm:TabItemViewModel">
          <UserControl Content="{Binding ViewModel}"/>
        </DataTemplate>
      </TabControl.ContentTemplate>
    </TabControl>
```

And here's my `MainWindowViewModel` and `TabItemViewModel` code:

```c#
    public class MainWindowViewModel : ViewModelBase
    {
        public ObservableCollection<TabItemViewModel> Tabs { get; }

        public MainWindowViewModel()
        {
            Tabs = new()
            {
                new TabItemViewModel("vm1", new ViewModel1() ),
                new TabItemViewModel("vm2", new ViewModel2() ),
                new TabItemViewModel("vm3", new ViewModel3() )
            };
        }
    }

    public class TabItemViewModel : ViewModelBase
    {
        private string _header;

        public string Header
        {
            get => _header;
            set => this.RaiseAndSetIfChanged(ref _header, value);
        }

        private ViewModelBase _viewModel;

        public ViewModelBase ViewModel
        {
            get => _viewModel;
            set => this.RaiseAndSetIfChanged(ref _viewModel, value);
        }
        
        public TabItemViewModel(string header, ViewModelBase viewModel)
        {
            Header = header;
            ViewModel = viewModel;
        }
    }
```

That way I can add any `ViewModelBase` to my `TabControl`, but they need to have corrseponding `View.axaml`s I guess. That's what I meant by "using one view": I'd like to have one `View.axaml` for each of these different classes, implementing `ViewModelBase`.

**Reply by jp2masa** [reply]
Instead of:

```xaml
          <UserControl Content="{Binding ViewModel}"/>
```

You can create the `View` directly:

```xaml
          <View DataContext="{Binding ViewModel}" />
```

**Reply by Gigas002** [reply]
Thank you, that's kind of working for me. I've experienced some minor issues though (not actually all of views in TabControl should be the same, just most part of them; some weird calls to methods with ViewModel's properties set to `null` without initializing it, etc), but I think I'll manage to resolve them.

**Reply by Gigas002** [reply]
In the end I decided to change `ViewLocator`'s `Build` method. I've modified the line to get type for `Activator.CreateInstance(type)` this way:

```c#
var type = paramType.IsSubclassOf(typeof(AbstractViewModel)) ? typeof(AbstractView) : Type.GetType(name);
```

Where `AbstractViewModel` is the parent `ViewModel` for all `ViewModels` that you want to have one view, specified here as `AbstractView`.
